### DISCUSSION TITLE
Fluctuating energy price support

### DISCUSSION THREAD

**Discussion by xoferz** [jmcollin78/versatile_thermostat#230] [Ideas]
Maybe it would be possible to introduce some kind or energy price support in the thermostat?

For instance, I am using a heat pump with underfloor heating. Electricity prices are freely available for me with ENTSO-e Transparency Platform integration (Nordpool) and I use nordpool_diff HACS integration which will give me a value of the current hour's price level over the next n hours on a scale of -1 to 1 (the highest hour's price is assigned -1 and the cheapest price 1 and the rest in between). I am currently using this -1 ... 1 value to switch all my Versatile thermostats between eco/comfort/boost mode, but it is not ideal.

My house has a very good thermal efficiency and very long inertia. Basically not much will happen if it is 0℃ outside and I do not heat for 8 hours. So depending on the outside temperature, I could heat only a couple of hours a day when it is the cheapest. I am struggling to automate the amount of hours needed to heat depending on the difference of inside/outside temperature.

**Comment by MBRetrofit** [comment]
We have here Voluntary price for the small consumer and is integrated in my HA. For now I simulate only with my boiler, very easy in Automations.

I build from Helpers a Toggle type input_boolean.heating_all switch for starting/stoping all together the Vtherm's from each room. I use a Vtherm for every room. I can add this switch besides my boiler to be actuated when the price is very low.

Building Trigger ID's:
https://www.youtube.com/watch?v=fE_MYcXYwMI
Building boolean toggle switch:
https://www.youtube.com/watch?v=sNmonuw4EHo&t=3s

``` alias: PVPC Cheap
description: ""
trigger:
  - type: value
    platform: device
    device_id: 646e4c52617da7cdf9a9b6621fb6ac2a
    entity_id: b4ac3a5fffbf79c68748162fdbdfcaf0
    domain: sensor
    below: 0.09
    id: Below
    above: 0
  - type: value
    platform: device
    device_id: 646e4c52617da7cdf9a9b6621fb6ac2a
    entity_id: b4ac3a5fffbf79c68748162fdbdfcaf0
    domain: sensor
    above: 0.091
    id: Above
condition: []
action:
  - choose:
      - conditions:
          - condition: trigger
            id:
              - Below
        sequence:
          - type: turn_on
            device_id: f6887127272c19820a63be0612c595e2
            entity_id: c42bf515681ad82b50b654a80e02e6c1
            domain: switch
      - conditions:
          - condition: trigger
            id:
              - Above
        sequence:
          - type: turn_off
            device_id: f6887127272c19820a63be0612c595e2
            entity_id: c42bf515681ad82b50b654a80e02e6c1
            domain: switch
mode: single
```

**Reply by jmcollin78** [reply]
Please @MBRetrofit , put your yaml code into <> (code balise)

**Reply by MBRetrofit** [reply]
Sorry, now is solved

**Reply by xoferz** [reply]
I did not really mean a binary on/off but rather a sliding scale depending on price level and heating energy requirement over time. The desired logic would be:

- calculate the total heating time over X hours required to achieve or keep desired temp. For instance, lets say with -1 outside temp, I need to heat at least 2 hour in every 12 hours. With -10 outside, I need 4 hours. It might be different in different rooms (per thermostat).

- find the best priced 2 hours and use them

Unfortunately this math on this platform is a bit over my head.

**Reply by jmcollin78** [reply]
> calculate the total heating time over X hours required to achieve or keep desired temp

If you have a formula for that, I'm curious.

At short term, I will not introduce price into VTherm. It is not in its responsability. Its job is to heat/regulate a room.

**Reply by xoferz** [reply]
I do not :-( 
I was hoping VTherm could take that from history.

**Reply by jmcollin78** [reply]
There is no history management in VTherm all is calculated in real time.

**Comment by MBRetrofit** [comment]
You can install a buffer tank ( boiler ) between heat pum and floor heating if you don't have it already. The heat pump should not be directly connected with the floor heating system and this should heat the buffer tank, using time as a leverage and a higher starting temperature.

The buffer tank from energy point of view is like a battery and you can "charge it" whenever the price is low and use it when price is high.

This way you can use the full power of the VTherm between tank and floor heating and a simple Automations for using the heat pump at full power when the price is low and at idle when the tank will only need to compensate.

**Reply by xoferz** [reply]
Yeah, there is a buffer tank but it is not enough (200L). The issue I describe, can be solved in software instead of installing a 1000L tank (for which I do not have room anyway). As I said, I am almost there just by switching VTherm between Eco/Comfort/Boost according to electricity price. The only thing missing is that it could heat even at less (and cheaper) hours than it is now.

**Reply by MBRetrofit** [reply]
Luckily for you, things are not like that.

Liters are just an atribute from the "Energy" entity...I mean from a floor heating point of you 1000 Liters tank is equivalent with the 200 Liters tanks if the latest receive a temperature increase. 
With a simple 3 way valve you will mix the tour and the return and the needed floor heating temperature agent will be avalable till price will go down again. 

No need to tremble when the price is high and also no need to overheat the floor, to consume the humidity and all rest.
